File: mobile_sdks/android/src/main/kotlin/com/example/dms_sdk/DMSClient.kt
package com.example.dms_sdk

import kotlinx.coroutines.*
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * DMS Android SDK - Main client for Document Management System
 * Version: 3.3.0
 */
class DMSClient private constructor() {

    companion object {
        @Volatile
        private var instance: DMSClient? = null

        fun getInstance(): DMSClient {
            return instance ?: synchronized(this) {
                instance ?: DMSClient().also { instance = it }
            }
        }
    }

    // Configuration
    data class Configuration(
        val baseURL: String,
        val apiKey: String? = null,
        val timeout: Long = 30L,
        val enableLogging: Boolean = false
    )

    private lateinit var configuration: Configuration
    private var apiKey: String? = null
    private var accessToken: String? = null

    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = true
    }

    private lateinit var client: OkHttpClient

    private fun initializeClient() {
        client = OkHttpClient.Builder()
            .connectTimeout(configuration.timeout, TimeUnit.SECONDS)
            .readTimeout(configuration.timeout, TimeUnit.SECONDS)
            .writeTimeout(configuration.timeout, TimeUnit.SECONDS)
            .apply {
                if (configuration.enableLogging) {
                    addInterceptor(HttpLoggingInterceptor().apply {
                        level = HttpLoggingInterceptor.Level.BODY
                    })
                }
            }
            .build()
    }

    /**
     * Configure the SDK
     */
    fun configure(config: Configuration) {
        this.configuration = config
        this.apiKey = config.apiKey
        initializeClient()

        if (config.enableLogging) {
            println("[DMS SDK] Configured with base URL: ${config.baseURL}")
        }
    }

    /**
     * Set API key for authentication
     */
    fun setAPIKey(key: String) {
        this.apiKey = key
    }

    /**
     * Set OAuth access token
     */
    fun setAccessToken(token: String) {
        this.accessToken = token
    }

    // MARK: - Authentication

    /**
     * Login with username and password
     */
    suspend fun login(username: String, password: String): Result<AuthResponse> {
        val body = mapOf(
            "username" to username,
            "password" to password
        )

        return try {
            val response = post<AuthResponse>("/api/v1/auth/login", body)
            response.onSuccess {
                accessToken = it.accessToken
            }
            response
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Logout current user
     */
    suspend fun logout(): Result<Unit> {
        return try {
            post<EmptyResponse>("/api/v1/auth/logout", null)
            accessToken = null
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK: - Documents

    /**
     * Get list of documents
     */
    suspend fun getDocuments(
        page: Int = 1,
        pageSize: Int = 20
    ): Result<DocumentListResponse> {
        val queryParams = mapOf(
            "page" to page.toString(),
            "page_size" to pageSize.toString()
        )
        return get("/api/v1/documents", queryParams)
    }

    /**
     * Get document by ID
     */
    suspend fun getDocument(id: String): Result<Document> {
        return get("/api/v1/documents/$id")
    }

    /**
     * Create new document
     */
    suspend fun createDocument(document: DocumentCreate): Result<Document> {
        return post("/api/v1/documents", document)
    }

    /**
     * Update existing document
     */
    suspend fun updateDocument(id: String, document: DocumentUpdate): Result<Document> {
        return put("/api/v1/documents/$id", document)
    }

    /**
     * Delete document
     */
    suspend fun deleteDocument(id: String): Result<Unit> {
        return try {
            delete<EmptyResponse>("/api/v1/documents/$id")
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Upload document file
     */
    suspend fun uploadDocument(
        file: File,
        metadata: Map<String, String>? = null
    ): Result<Document> {
        return withContext(Dispatchers.IO) {
            try {
                val requestBody = MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart(
                        "file",
                        file.name,
                        file.asRequestBody("application/octet-stream".toMediaType())
                    )

                metadata?.forEach { (key, value) ->
                    requestBody.addFormDataPart(key, value)
                }

                val request = Request.Builder()
                    .url("${configuration.baseURL}/api/v1/documents/upload")
                    .post(requestBody.build())
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<Document>(response)
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Upload failed"))
            }
        }
    }

    /**
     * Download document file
     */
    suspend fun downloadDocument(id: String): Result<ByteArray> {
        return withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("${configuration.baseURL}/api/v1/documents/$id/download")
                    .get()
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()

                if (!response.isSuccessful) {
                    Result.failure(DMSException.HttpError(response.code))
                } else {
                    Result.success(response.body?.bytes() ?: ByteArray(0))
                }
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Download failed"))
            }
        }
    }

    // MARK: - Search

    /**
     * Search documents
     */
    suspend fun searchDocuments(
        query: String,
        filters: Map<String, Any>? = null
    ): Result<DocumentListResponse> {
        val body = mutableMapOf<String, Any>("query" to query)
        filters?.let { body["filters"] = it }

        return post("/api/v1/search", body)
    }

    // MARK: - Analytics

    /**
     * Get dashboard analytics
     */
    suspend fun getDashboard(): Result<DashboardResponse> {
        return get("/api/v1/analytics/dashboard")
    }

    /**
     * Get document statistics
     */
    suspend fun getStatistics(
        startDate: String? = null,
        endDate: String? = null
    ): Result<StatisticsResponse> {
        val queryParams = mutableMapOf<String, String>()
        startDate?.let { queryParams["start_date"] = it }
        endDate?.let { queryParams["end_date"] = it }

        return get("/api/v1/analytics/statistics", queryParams)
    }

    // MARK: - Notifications

    /**
     * Get notifications
     */
    suspend fun getNotifications(
        page: Int = 1,
        pageSize: Int = 20
    ): Result<NotificationListResponse> {
        val queryParams = mapOf(
            "page" to page.toString(),
            "page_size" to pageSize.toString()
        )
        return get("/api/v1/notifications", queryParams)
    }

    /**
     * Mark notification as read
     */
    suspend fun markNotificationAsRead(id: String): Result<Unit> {
        return try {
            post<EmptyResponse>("/api/v1/notifications/$id/read", null)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    // MARK: - WebSocket (Real-time)

    /**
     * Connect to WebSocket for real-time updates
     */
    fun connectWebSocket(
        onMessage: (WebSocketMessage) -> Unit,
        onError: (Throwable) -> Unit,
        onClosed: (Int, String) -> Unit
    ): WebSocket {
        val wsUrl = configuration.baseURL
            .replace("https://", "wss://")
            .replace("http://", "ws://") + "/ws"

        val request = Request.Builder()
            .url(wsUrl)
            .apply { addAuthHeaders() }
            .build()

        return client.newWebSocket(request, object : WebSocketListener() {
            override fun onMessage(webSocket: WebSocket, text: String) {
                try {
                    val message = json.decodeFromString<WebSocketMessage>(text)
                    onMessage(message)
                } catch (e: Exception) {
                    onError(e)
                }
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                onError(t)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                onClosed(code, reason)
            }
        })
    }

    // MARK: - Private Helpers

    private suspend inline fun <reified T> get(
        endpoint: String,
        queryParams: Map<String, String>? = null
    ): Result<T> {
        return withContext(Dispatchers.IO) {
            try {
                val urlBuilder = HttpUrl.parse("${configuration.baseURL}$endpoint")!!.newBuilder()
                queryParams?.forEach { (key, value) ->
                    urlBuilder.addQueryParameter(key, value)
                }

                val request = Request.Builder()
                    .url(urlBuilder.build())
                    .get()
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<T>(response)
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Request failed"))
            }
        }
    }

    private suspend inline fun <reified T> post(
        endpoint: String,
        body: Any?
    ): Result<T> {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = body?.let {
                    json.encodeToString(it).toRequestBody("application/json".toMediaType())
                } ?: "{}".toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url("${configuration.baseURL}$endpoint")
                    .post(jsonBody)
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<T>(response)
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Request failed"))
            }
        }
    }

    private suspend inline fun <reified T> put(
        endpoint: String,
        body: Any
    ): Result<T> {
        return withContext(Dispatchers.IO) {
            try {
                val jsonBody = json.encodeToString(body).toRequestBody("application/json".toMediaType())

                val request = Request.Builder()
                    .url("${configuration.baseURL}$endpoint")
                    .put(jsonBody)
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<T>(response)
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Request failed"))
            }
        }
    }

    private suspend inline fun <reified T> delete(endpoint: String): Result<T> {
        return withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("${configuration.baseURL}$endpoint")
                    .delete()
                    .apply { addAuthHeaders() }
                    .build()

                val response = client.newCall(request).execute()
                handleResponse<T>(response)
            } catch (e: Exception) {
                Result.failure(DMSException.NetworkError(e.message ?: "Request failed"))
            }
        }
    }

    private fun Request.Builder.addAuthHeaders() {
        addHeader("Content-Type", "application/json")
        addHeader("Accept", "application/json")

        apiKey?.let { addHeader("X-API-Key", it) }
        accessToken?.let { addHeader("Authorization", "Bearer $it") }
    }

    private inline fun <reified T> handleResponse(response: Response): Result<T> {
        return try {
            if (!response.isSuccessful) {
                val errorBody = response.body?.string()
                val errorMessage = try {
                    val error = json.decodeFromString<ErrorResponse>(errorBody ?: "{}")
                    error.message
                } catch (e: Exception) {
                    "HTTP ${response.code}"
                }
                Result.failure(DMSException.ApiError(errorMessage))
            } else {
                val body = response.body?.string() ?: "{}"
                val result = json.decodeFromString<T>(body)
                Result.success(result)
            }
        } catch (e: Exception) {
            Result.failure(DMSException.DecodingError(e.message ?: "Decoding failed"))
        }
    }
}

// MARK: - Exceptions

sealed class DMSException(message: String) : Exception(message) {
    class NetworkError(message: String) : DMSException(message)
    class HttpError(val statusCode: Int) : DMSException("HTTP error: $statusCode")
    class ApiError(message: String) : DMSException(message)
    class DecodingError(message: String) : DMSException("Decoding error: $message")
    class InvalidURL : DMSException("Invalid URL")
    class Unauthorized : DMSException("Unauthorized access")
    class NotFound : DMSException("Resource not found")
}

// Empty response for operations that don't return data
@Serializable
private class EmptyResponse

@Serializable
private data class ErrorResponse(
    val message: String,
    val code: String? = null
)

File: mobile_sdks/react_native/src/DMSClient.ts
/**
 * DMS React Native SDK - Main client for Document Management System
 * Version: 3.3.0
 */

import {
  AuthResponse,
  User,
  Document,
  DocumentCreate,
  DocumentUpdate,
  DocumentListResponse,
  DashboardResponse,
  StatisticsResponse,
  NotificationListResponse,
  WebSocketMessage,
  DMSError,
  Configuration,
} from './types';

export class DMSClient {
  private static instance: DMSClient;
  private baseURL: string = 'https://api.dms.example.com';
  private apiKey?: string;
  private accessToken?: string;
  private config: Configuration;
  private ws?: WebSocket;

  private constructor() {
    this.config = {
      baseURL: 'https://api.dms.example.com',
      timeout: 30000,
      enableLogging: false,
    };
  }

  /**
   * Get singleton instance
   */
  public static getInstance(): DMSClient {
    if (!DMSClient.instance) {
      DMSClient.instance = new DMSClient();
    }
    return DMSClient.instance;
  }

  /**
   * Configure the SDK
   */
  public configure(config: Configuration): void {
    this.config = { ...this.config, ...config };
    this.baseURL = config.baseURL;
    this.apiKey = config.apiKey;

    if (config.enableLogging) {
      console.log(`[DMS SDK] Configured with base URL: ${config.baseURL}`);
    }
  }

  /**
   * Set API key for authentication
   */
  public setAPIKey(key: string): void {
    this.apiKey = key;
  }

  /**
   * Set OAuth access token
   */
  public setAccessToken(token: string): void {
    this.accessToken = token;
  }

  // MARK: - Authentication

  /**
   * Login with username and password
   */
  public async login(username: string, password: string): Promise<AuthResponse> {
    const response = await this.request<AuthResponse>('POST', '/api/v1/auth/login', {
      username,
      password,
    });

    this.accessToken = response.accessToken;
    return response;
  }

  /**
   * Logout current user
   */
  public async logout(): Promise<void> {
    await this.request('POST', '/api/v1/auth/logout');
    this.accessToken = undefined;
  }

  // MARK: - Documents

  /**
   * Get list of documents
   */
  public async getDocuments(page: number = 1, pageSize: number = 20): Promise<DocumentListResponse> {
    return this.request<DocumentListResponse>('GET', '/api/v1/documents', undefined, {
      page: page.toString(),
      page_size: pageSize.toString(),
    });
  }

  /**
   * Get document by ID
   */
  public async getDocument(id: string): Promise<Document> {
    return this.request<Document>('GET', `/api/v1/documents/${id}`);
  }

  /**
   * Create new document
   */
  public async createDocument(document: DocumentCreate): Promise<Document> {
    return this.request<Document>('POST', '/api/v1/documents', document);
  }

  /**
   * Update existing document
   */
  public async updateDocument(id: string, document: DocumentUpdate): Promise<Document> {
    return this.request<Document>('PUT', `/api/v1/documents/${id}`, document);
  }

  /**
   * Delete document
   */
  public async deleteDocument(id: string): Promise<void> {
    await this.request('DELETE', `/api/v1/documents/${id}`);
  }

  /**
   * Upload document file
   */
  public async uploadDocument(
    file: File | Blob,
    filename: string,
    metadata?: Record<string, string>
  ): Promise<Document> {
    const formData = new FormData();
    formData.append('file', file, filename);

    if (metadata) {
      Object.entries(metadata).forEach(([key, value]) => {
        formData.append(key, value);
      });
    }

    const url = `${this.baseURL}/api/v1/documents/upload`;
    const headers: HeadersInit = {};

    this.addAuthHeaders(headers);

    const response = await fetch(url, {
      method: 'POST',
      headers,
      body: formData,
    });

    if (!response.ok) {
      throw await this.handleErrorResponse(response);
    }

    return response.json();
  }

  /**
   * Download document file
   */
  public async downloadDocument(id: string): Promise<Blob> {
    const url = `${this.baseURL}/api/v1/documents/${id}/download`;
    const headers: HeadersInit = {};

    this.addAuthHeaders(headers);

    const response = await fetch(url, {
      method: 'GET',
      headers,
    });

    if (!response.ok) {
      throw await this.handleErrorResponse(response);
    }

    return response.blob();
  }

  // MARK: - Search

  /**
   * Search documents
   */
  public async searchDocuments(
    query: string,
    filters?: Record<string, any>
  ): Promise<DocumentListResponse> {
    return this.request<DocumentListResponse>('POST', '/api/v1/search', {
      query,
      filters,
    });
  }

  // MARK: - Analytics

  /**
   * Get dashboard analytics
   */
  public async getDashboard(): Promise<DashboardResponse> {
    return this.request<DashboardResponse>('GET', '/api/v1/analytics/dashboard');
  }

  /**
   * Get document statistics
   */
  public async getStatistics(startDate?: string, endDate?: string): Promise<StatisticsResponse> {
    const queryParams: Record<string, string> = {};
    if (startDate) queryParams.start_date = startDate;
    if (endDate) queryParams.end_date = endDate;

    return this.request<StatisticsResponse>('GET', '/api/v1/analytics/statistics', undefined, queryParams);
  }

  // MARK: - Notifications

  /**
   * Get notifications
   */
  public async getNotifications(page: number = 1, pageSize: number = 20): Promise<NotificationListResponse> {
    return this.request<NotificationListResponse>('GET', '/api/v1/notifications', undefined, {
      page: page.toString(),
      page_size: pageSize.toString(),
    });
  }

  /**
   * Mark notification as read
   */
  public async markNotificationAsRead(id: string): Promise<void> {
    await this.request('POST', `/api/v1/notifications/${id}/read`);
  }

  // MARK: - WebSocket (Real-time)

  /**
   * Connect to WebSocket for real-time updates
   */
  public connectWebSocket(
    onMessage: (message: WebSocketMessage) => void,
    onError: (error: Event) => void,
    onClose: (event: CloseEvent) => void
  ): WebSocket {
    const wsUrl = this.baseURL
      .replace('https://', 'wss://')
      .replace('http://', 'ws://') + '/ws';

    this.ws = new WebSocket(wsUrl);

    this.ws.onopen = () => {
      if (this.config.enableLogging) {
        console.log('[DMS SDK] WebSocket connected');
      }
    };

    this.ws.onmessage = (event: MessageEvent) => {
      try {
        const message: WebSocketMessage = JSON.parse(event.data);
        onMessage(message);
      } catch (error) {
        console.error('[DMS SDK] Failed to parse WebSocket message:', error);
      }
    };

    this.ws.onerror = (error: Event) => {
      if (this.config.enableLogging) {
        console.error('[DMS SDK] WebSocket error:', error);
      }
      onError(error);
    };

    this.ws.onclose = (event: CloseEvent) => {
      if (this.config.enableLogging) {
        console.log('[DMS SDK] WebSocket closed');
      }
      onClose(event);
    };

    return this.ws;
  }

  /**
   * Disconnect WebSocket
   */
  public disconnectWebSocket(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = undefined;
    }
  }

  // MARK: - Private Helpers

  private async request<T>(
    method: string,
    endpoint: string,
    body?: any,
    queryParams?: Record<string, string>
  ): Promise<T> {
    let url = `${this.baseURL}${endpoint}`;

    if (queryParams) {
      const params = new URLSearchParams(queryParams);
      url += `?${params.toString()}`;
    }

    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    this.addAuthHeaders(headers);

    if (this.config.enableLogging) {
      console.log(`[DMS SDK] ${method} ${url}`);
    }

    const response = await fetch(url, {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    });

    if (!response.ok) {
      throw await this.handleErrorResponse(response);
    }

    // For DELETE requests, return empty response
    if (method === 'DELETE') {
      return {} as T;
    }

    return response.json();
  }

  private addAuthHeaders(headers: HeadersInit): void {
    if (this.apiKey) {
      headers['X-API-Key'] = this.apiKey;
    }

    if (this.accessToken) {
      headers['Authorization'] = `Bearer ${this.accessToken}`;
    }
  }

  private async handleErrorResponse(response: Response): Promise<DMSError> {
    let message = `HTTP ${response.status}`;

    try {
      const errorData = await response.json();
      message = errorData.message || message;
    } catch (e) {
      // Failed to parse error response
    }

    return new DMSError(
      message,
      response.status,
      response.status === 401 ? 'unauthorized' :
      response.status === 404 ? 'not_found' :
      response.status >= 500 ? 'server_error' :
      'api_error'
    );
  }
}

export default DMSClient;
